python/ray/data/_internal/pipeline_executor.py

```diff
@@ -19,7 +19,9 @@ def pipeline_stage(fn: Callable[[], Dataset[T]]) -> Dataset[T]:
     # Force eager evaluation of all blocks in the pipeline stage. This
     # prevents resource deadlocks due to overlapping stage execution (e.g.,
     # task -> actor stage).
-    return fn().fully_executed()
+    ds = fn()
+    ds._plan.execute(force_read=True)
```

Comment:
Isn't this the same as fully_executed()? Do we not want to cache the output blocks here?

Reply:
The output blocks of transformations before the repeat/window are cached, but the outputs of pipeline are not. E.g. `ray.data.read_parquet().map_batches().window().map_batches().repeat(N)`, the results up to the first `map_batches()` are cached and reused, but the output of the second `map_batches()` are not.

Also I think the output blocks here by `pipeline_stage()` are not even the final outputs from the pipeline yet.

Reply:
Got it. Could we add a comment here to this effect?

Reply:
Yep, will do and polish once tests are fixed.